transport/client.go

```diff
@@ -0,0 +1,239 @@
+package transport
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"net"
+	"os"
+	"time"
+
+	"github.com/cloudflare/cfssl/csr"
+	"github.com/cloudflare/cfssl/log"
+	"github.com/cloudflare/cfssl/transport/ca"
+	"github.com/cloudflare/cfssl/transport/core"
+	"github.com/cloudflare/cfssl/transport/kp"
+)
+
+func envOrDefault(key, def string) string {
+	val := os.Getenv(key)
+	if val == "" {
+		return def
+	}
+	return val
+}
+
+var (
+	// NewKeyProvider is the function used to build key providers
+	// from some identity.
+	NewKeyProvider = func(id *core.Identity) (kp.KeyProvider, error) {
+		return kp.NewStandardProvider(id)
+	}
+
+	// NewCA is used to load a configuration for a certificate
+	// authority.
+	NewCA = func(id *core.Identity) (ca.CertificateAuthority, error) {
+		return ca.NewCFSSLProvider(id, nil)
+	}
+)
+
+// A Transport is capable of providing transport-layer security using
+// TLS.
+type Transport struct {
+	// Before defines how long before the certificate expires the
+	// transport should start attempting to refresh the
+	// certificate. For example, if this is 24h, then 24 hours
+	// before the certificate expires the Transport will start
+	// attempting to replace it.
+	Before time.Duration
+
+	// Provider contains a key management provider.
+	Provider kp.KeyProvider
+
+	// CA contains a mechanism for obtaining signed certificates.
+	CA ca.CertificateAuthority
+
+	// Roots contains the pool of X.509 certificates that are
+	// valid for authenticating peers.
+	Roots *x509.CertPool
+
+	// Identity contains information about the entity that will be
+	// used to construct certificates.
+	Identity *core.Identity
+}
+
+// NewTransport builds a new transport from the default
+func NewTransport(before time.Duration, identity *core.Identity) (*Transport, error) {
+	var tr = &Transport{
+		Before:   before,
+		Roots:    core.SystemRoots,
```

Comment:
Something is missing here. This doesn't seem to work in the use case where two services, backed by two different CFSSL CAs attempt to talk with each other. Unless the CFSSL CAs were previously added to the system roots, the CFSSL-signed certificates will not be trusted by the client.

When you build a transport, you should be able to choose a set of CA providers to trust rather than trust the system roots. By providing a list of CAs to trust, the transport package could use the CACertificate() function to obtain the trusted roots of those certificates and create a very specialized trust store.

The use case I'm envisioning here is represented by these images:
https://blog.cloudflare.com/content/images/2015/06/provisioning-app-certificates.png https://blog.cloudflare.com/content/images/2015/06/service-to-service-communication.png